<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>基本用法</title>
</head>
<body>
    <script>
    //ES6规定,Promise对象是一个构造函数,用来生成Promise实例
    //下面的代码创造了一个Promise实例
/*    var promise=new Promise(functon(resolve,reject){
        //...some code
    });
    if(异步操作成功){
        resolve(value)
    }else{
        reject(error)
    }
*/
    //Promise构造函数接受一个函数作为参数,该函数的两个参数分别是resolve和reject他们是两个函数,由JavaScript引擎提供,不用自己部署
    //resolve函数的作用是,将Promise对象的状态从'未完成'变成'成功'(即从Pending变为Resolved),在异步操作成功时调用,并将异步操作的结果作为参数传递出去;
    //reject函数的作用是,将Promise对象的状态从'未完成'到'失败'(即Pending变为Rejected),在异步操作失败时调用,并将异步操作报出的错误作为参数传递出去

    //Promise实例生成后,可以用then方法分别制定Resolved状态和Rejected状态的回调函数
/*
    promise.then(function(value){
        //success
    },function(error){
        //failure
    })
*/
    //then方法可以接受两个回调函数作为参数.
    //第一个回调函数是Promise对象的状态变为Resolved时调用,第二个函数是Promise对象的状态变为Rejected时调用.
    //其中,第二个函数是可选的,不一定要提供.
    //这两个函数都接受Promise对象传出的值作为参数

    //下面是一个Promise对象的简单例子
    function timeout(ms){
        return new Promise((resolve,reject)=>{
            setTimeout(resolve,ms,'done');
        })
    }
    timeout(1000).then((value)=>{
        console.log(value);
    })

    //上面的代码中,timeout方法返回一个Promise实例,表示一段时间以后会发生的结果.
    //过了指定的时间(ms参数)以后,Promise实例的状态变为Resolved,就会出发then方法绑定的回调函数

    //Promise新建后就会立即执行
    let promise=new Promise((resolve,reject)=>{
        console.log('Promise');
        resolve();
    });
    promise.then(function(){
        console.log('Resolved');
    });
    console.log('Hi!');

    //Promise Hi Resolved

    //上面的代码中,Promise新建后会立即执行,所以首先输出的是Promise.
    //然后,then方法指定的回调函数将在当前脚本所有同步任务执行完成后才会执行,所以Resolved最后输出

    //下面是异步加载图片的例子
    function loadImageAsync(url){
        return new Promise(function(resolve,reject){
            var image=new Image();
            image.onload=function(){
                resolve(image);
            };
            image.onerror=function(){
                reject(new Error('Could not load image at '+url));
            };

            image.src=url; 
        });
    }

    //上面的代码中使用Promise包装了一个图片加载的异步操作.
    //如果加载成功,就调用resolve方法,否则就调用reject方法

    //下面是一个用Promise对象实现的AJAX操作的例子
    var getJSON=function(url){
        var promise=new Promise(function(resolve,reject){
            var client=new XMLHttpRequest();
            client.open('GET',url,true);
            client.responseType='json';
            client.setRequestHeader('Accept','application/json');
            client.send(null);
            client.onreadystatechange=function(){
                if(this.readyState!==4) return;
                if(this.status===200){
                    resolve(this.response);
                }else{
                    reject(new Error(this.statusText));
                }
            }
        })
        return promise;
    };
    
    // getJSON('/posts.json').then(function(json){
    //     console.log('Contents: '+json);
    // },function(error){
    //     console.log('出错了',error);
    // });

    //上面的代码中,getJSON是对XMLHttpRequest对象的封装,用于发出一个针对JSON数据的HTTP请求,并返回一个Promise对象,
    //需要注意的是,在getJSON内部,resolve函数和reject函数调用时都带有参数

    //如果调用resolve函数和reject函数时都带有参数,那么这些参数会被传递给回调函数.
    //reject函数的参数通常是Error对象的实例,表示抛出的错误;
    //resolve函数的参数除了正常的值外,还可能是另一个Promise实例,比如下面这样.

    // var p1=new Promise(function(resolve,reject){
    //     //...
    // })

    // var p2=new Promise(function(resolve,reject){
    //     resolve(p1);
    // })
    //上面的代码中,p1和p2都是Promise的实例,但是p2的resolve方法将p1作为参数,即一个异步操作的结果是返回另一个异步操作
    
    /*
    ***注意:此时的p1的状态就会传递给p2.也就是说,p1的状态决定了p2的状态.
            如果p1的状态是Pending,那么p2的回调函数就会等待p1的状态改变,
            如果p1的状态已经是Resolved或Rejected,那么p2的回调函数将会立即执行
    */

    var p1=new Promise(function(resolve,reject){
        setTimeout(()=>reject(new Error('fail')),3000)
    })
    var p2=new Promise(function(resolve,reject){
        setTimeout(()=>resolve(p1),1000);
    })

    p2.then(result=>console.log(result)).catch(error=>console.log(error));

    //上面的代码中,p1是一个Promise,3秒之后变为rejected.
    //p2的状态在1秒之后改变,resolve方法返回的是p1.
    //由于p2返回的是另一个Promise,导致p2的状态无效,由p1的状态决定p2的状态
    //所以,后面的then语句就变成针对后者(p1)的.
    //再过2秒,p1就变为rejected,触发catch方法指定的回调函数

    //注意,调用resolve或reject并不会终结Promise的参数函数的执行
    new Promise((resolve,reject)=>{
        resovle(1);
        console.log(2);
    }).then(r=>{
        console.log(r);
    })
    //2
    //1

    //上面的代码中,调用resolve(1)以后,后面的console.log(2)还是会执行,并且会首先打印出来
    //这是因为立即resolved的Promise是在本轮时间UN换的末尾执行,总晚于本轮循环的同步任务
    //一般来说,调用resolve或reject以后,Promise的使命就完成了,后继操作应该放到then方法里面,而不应该直接写在resolve或reject后面
    //所以,最好在它们前面加上return语句,这样就不会产生意外

    new Promise((resolve,reject)=>{
        return resolve(1);
        //后面的语句不会执行
        console.log(2);
    })
    
    </script>
</body>
</html>